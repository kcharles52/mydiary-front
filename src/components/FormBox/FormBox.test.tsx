// react libraries
import * as React from "react";

// third-party libraries
import { mount, shallow } from "enzyme";

// components
import { mapDispatchToProps, mapStateToProps, FormBox } from "./";

describe("Formbox component", () => {
  let wrapper, shallowWrapper, props;
  props = {
    loginUserAction: jest.fn(() => Promise.resolve()),
    registerUserAction: jest.fn(() => Promise.resolve()),
    ToggleTab: jest.fn(),
    HandleInputChange: jest.fn(),
    history: { push: jest.fn() }
  };

  beforeEach(() => {
    wrapper = mount(<FormBox {...props} />);
    shallowWrapper = shallow(<FormBox {...props} />);
  });
  afterEach(() => {
    wrapper = shallowWrapper = null;
  });
  it("should match snapshot", () => {
    expect(shallowWrapper).toMatchSnapshot();
  });

  it("calls onChange prop when input value changes", () => {
    const spy = jest.spyOn(wrapper.instance(), "HandleInputChange");
    wrapper.instance().forceUpdate();
    wrapper
      .find("input")
      .first()
      .simulate("change");
    expect(spy).toHaveBeenCalled();
    spy.mockRestore();
  });

  it("calls  HandleSubmit prop when a submit button is clicked [login]", () => {
    const spy = jest.spyOn(wrapper.instance(), "HandleSubmit");
    wrapper.instance().forceUpdate();
    wrapper.setState({ user: {} });
    wrapper
      .find("Button[label='Login']")
      .first()
      .simulate("click");
    expect(spy).toHaveBeenCalled();
    spy.mockRestore();
  });
  it("calls  HandleSubmit prop when a submit button is clicked [register]", () => {
    const spy = jest.spyOn(wrapper.instance(), "HandleSubmit");
    wrapper.instance().forceUpdate();
    wrapper.setState({ user: {}, Signup: true });
    wrapper
      .find("Button[label='Get started']")
      .last()
      .simulate("click");
    expect(spy).toHaveBeenCalled();
    spy.mockRestore();
  });

  it("Should redirect to home page on login", () => {
    wrapper.setProps({
      user: { isLoggedIn: true, user: { token: "token string" } }
    });
    expect(props.history.push).toHaveBeenCalled();
  });

  describe("should change form", () => {
    let spy;
    beforeEach(() => {
      spy = jest.spyOn(wrapper.instance(), "ToggleTab");
    });

    afterEach(() => {
      spy.mockRestore();
    });

    it("Change to Register form", () => {
      wrapper
        .find(".tab")
        .last()
        .simulate("click");
      expect(spy).toHaveBeenLastCalledWith(true);
    });
    it("Change to login form", () => {
      wrapper
        .find(".tab")
        .first()
        .simulate("click");
      expect(wrapper.instance().ToggleTab).toHaveBeenLastCalledWith(false);
    });
  });
});


describe("Connect functions function", () => {
  describe("The mapStateToProps function", () => {
    it("should return the expected props object", () => {
      const state = {
        user: { name: "test", email: "test@email.com" }
      };
      const props = mapStateToProps(state);
      expect(props.user).toEqual(state.user);
    });
  });

  describe("The mapDispatchToProps function", () => {
    let dispatch, props;
    beforeEach(() => {
      dispatch = jest.fn();
      props = mapDispatchToProps(dispatch);
    });
    it("should dispatch loginUser when loginUserAction is called", () => {
      props.loginUserAction();
      expect(dispatch).toHaveBeenCalled();
    });
    it("should dispatch registerUser when registerUserAction is called", () => {
      props.registerUserAction();
      expect(dispatch).toHaveBeenCalled();
    });
  });
});
