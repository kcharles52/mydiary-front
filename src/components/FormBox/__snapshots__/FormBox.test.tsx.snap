// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Formbox component should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FormBox
    HandleInputChange={[MockFunction]}
    ToggleTab={[MockFunction]}
    history={
      Object {
        "push": [MockFunction],
      }
    }
    loginUserAction={[MockFunction]}
    registerUserAction={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <div
          className="tab-group"
        >
          <div
            className="tab"
            onClick={[Function]}
          >
            <a
              href="#login"
            >
              Log In
            </a>
          </div>
          <div
            className="tab"
            onClick={[Function]}
          >
            <a
              href="#signup"
            >
              Register
            </a>
          </div>
        </div>,
        <div
          className="tab-content"
        >
          <div>
            <LoginForm
              email=""
              name=""
              onChange={[Function]}
              onSubmit={[Function]}
            />
          </div>
          <div />
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="tab"
              onClick={[Function]}
            >
              <a
                href="#login"
              >
                Log In
              </a>
            </div>,
            <div
              className="tab"
              onClick={[Function]}
            >
              <a
                href="#signup"
              >
                Register
              </a>
            </div>,
          ],
          "className": "tab-group",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <a
                href="#login"
              >
                Log In
              </a>,
              "className": "tab",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Log In",
                "href": "#login",
              },
              "ref": null,
              "rendered": "Log In",
              "type": "a",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <a
                href="#signup"
              >
                Register
              </a>,
              "className": "tab",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Register",
                "href": "#signup",
              },
              "ref": null,
              "rendered": "Register",
              "type": "a",
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div>
              <LoginForm
                email=""
                name=""
                onChange={[Function]}
                onSubmit={[Function]}
              />
            </div>,
            <div />,
          ],
          "className": "tab-content",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <LoginForm
                email=""
                name=""
                onChange={[Function]}
                onSubmit={[Function]}
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "email": "",
                "name": "",
                "onChange": [Function],
                "onSubmit": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <div
            className="tab-group"
          >
            <div
              className="tab"
              onClick={[Function]}
            >
              <a
                href="#login"
              >
                Log In
              </a>
            </div>
            <div
              className="tab"
              onClick={[Function]}
            >
              <a
                href="#signup"
              >
                Register
              </a>
            </div>
          </div>,
          <div
            className="tab-content"
          >
            <div>
              <LoginForm
                email=""
                name=""
                onChange={[Function]}
                onSubmit={[Function]}
              />
            </div>
            <div />
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="tab"
                onClick={[Function]}
              >
                <a
                  href="#login"
                >
                  Log In
                </a>
              </div>,
              <div
                className="tab"
                onClick={[Function]}
              >
                <a
                  href="#signup"
                >
                  Register
                </a>
              </div>,
            ],
            "className": "tab-group",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <a
                  href="#login"
                >
                  Log In
                </a>,
                "className": "tab",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Log In",
                  "href": "#login",
                },
                "ref": null,
                "rendered": "Log In",
                "type": "a",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <a
                  href="#signup"
                >
                  Register
                </a>,
                "className": "tab",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Register",
                  "href": "#signup",
                },
                "ref": null,
                "rendered": "Register",
                "type": "a",
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div>
                <LoginForm
                  email=""
                  name=""
                  onChange={[Function]}
                  onSubmit={[Function]}
                />
              </div>,
              <div />,
            ],
            "className": "tab-content",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <LoginForm
                  email=""
                  name=""
                  onChange={[Function]}
                  onSubmit={[Function]}
                />,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "email": "",
                  "name": "",
                  "onChange": [Function],
                  "onSubmit": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
